File: utilsLibrary/src/main/java/com/zeropercenthappy/utilslibrary/FileUtils.java
package com.zeropercenthappy.utilslibrary;

import android.text.TextUtils;
import android.webkit.MimeTypeMap;

import java.io.BufferedOutputStream;
import java.io.BufferedWriter;
import java.io.Closeable;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;

/**
 * @author ybq
 */

public class FileUtils {
    /**
     * io流缓存大小
     */
    private static final int BUFFER_SIZE = 8192;

    /**
     * 关闭io流
     *
     * @param closeables
     */
    public static void closeIO(Closeable... closeables) {
        if (null == closeables || closeables.length <= 0) {
            return;
        }
        for (Closeable cb : closeables) {
            try {
                if (null == cb) {
                    continue;
                }
                cb.close();
            } catch (IOException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * 检查文件是否存在，若不存在则新建文件
     *
     * @param file
     * @return
     */
    public static boolean checkFileAndCreate(File file) {
        if (file == null) {
            return false;
        }
        if (!file.exists()) {
            try {
                return file.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
                return false;
            }
        } else {
            return true;
        }
    }

    /**
     * 检查文件是否存在，若不存在则新建文件
     *
     * @param path
     * @return
     */
    public static boolean checkFileAndCreate(String path) {
        if (TextUtils.isEmpty(path)) {
            return false;
        }
        File file = new File(path);
        if (!file.exists()) {
            try {
                return file.createNewFile();
            } catch (IOException e) {
                e.printStackTrace();
                return false;
            }
        } else {
            return true;
        }
    }

    /**
     * 通过InputStream写入文件
     */
    public static boolean writeFileByIS(File storageFile, InputStream inputStream, boolean append) {
        if (!checkFileAndCreate(storageFile) || inputStream == null) {
            return false;
        }
        OutputStream outputStream = null;
        try {
            outputStream = new BufferedOutputStream(new FileOutputStream(storageFile, append));
            byte[] data = new byte[BUFFER_SIZE];
            int len;
            while ((len = inputStream.read(data, 0, BUFFER_SIZE)) != -1) {
                outputStream.write(data, 0, len);
            }
            return true;
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        } finally {
            closeIO(inputStream, outputStream);
        }
    }

    /**
     * 向文件写入content
     *
     * @param filename 目标文件
     * @param content  内容
     * @param append   是否追加
     * @return
     */
    public static boolean writeContent2File(String filename, String content, boolean append) {
        boolean isSuccess = false;
        BufferedWriter bufferedWriter = null;
        try {
            bufferedWriter = new BufferedWriter(new FileWriter(filename, append));
            bufferedWriter.write(content);
            isSuccess = true;
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            closeIO(bufferedWriter);
        }
        return isSuccess;
    }

    /**
     * 获取文件的mimeType
     *
     * @param file
     * @return
     */
    public static String getFileMimeType(File file) {
        String path = file.getAbsolutePath();

        if (TextUtils.isEmpty(path)) {
            return null;
        }

        int dotPosition = path.lastIndexOf(".");
        if (dotPosition == -1) {
            return null;
        }
        String extension = path.substring(dotPosition + 1);
        extension = extension.toLowerCase();
        return MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
    }

    /**
     * 获取文件的mimeType
     *
     * @param path
     * @return
     */
    public static String getFileMimeType(String path) {
        if (TextUtils.isEmpty(path)) {
            return null;
        }

        int dotPosition = path.lastIndexOf(".");
        if (dotPosition == -1) {
            return null;
        }
        String extension = path.substring(dotPosition + 1);
        extension = extension.toLowerCase();
        return MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension);
    }

    /**
     * 删除文件或目录
     *
     * @param file
     * @return
     */
    public static boolean deleteFile(File file) {
        if (file == null || !file.exists()) {
            return false;
        }

        if (file.isFile()) {
            return file.delete();
        } else if (file.isDirectory()) {
            File[] files = file.listFiles();
            for (File subFile : files) {
                deleteFile(subFile);
            }
            return file.delete();
        } else {
            return false;
        }
    }

    /**
     * 获取文件或目录的大小，单位：Byte
     *
     * @param file
     * @return
     */
    public static long getSize(File file) {
        long size = 0;

        if (file == null || !file.exists()) {
            return size;
        }
        if (file.isFile()) {
            return file.length();
        } else if (file.isDirectory()) {
            File[] files = file.listFiles();
            for (File subFile : files) {
                size += getSize(subFile);
            }
        }

        return size;
    }

    /**
     * 获取文件或目录的大小，单位：Byte
     *
     * @param path
     * @return
     */
    public static long getSize(String path) {
        if (TextUtils.isEmpty(path)) {
            return 0;
        }
        File file = new File(path);
        long size = 0;

        if (!file.exists()) {
            return size;
        }
        if (file.isFile()) {
            return file.length();
        } else if (file.isDirectory()) {
            File[] files = file.listFiles();
            for (File subFile : files) {
                size += getSize(subFile);
            }
        }

        return size;
    }

    /**
     * 复制文件
     *
     * @param sourceFile  源文件
     * @param tartgetFile 目标文件
     * @param replace     是否替换
     * @return
     */
    public static boolean copyFile(File sourceFile, File tartgetFile, boolean replace) {
        if (sourceFile == null || !sourceFile.exists()) {
            return false;
        }
        if (replace && tartgetFile != null && tartgetFile.exists()) {
            if (!deleteFile(tartgetFile)) {
                //若目标文件删除失败
                return false;
            }
        }
        if (!checkFileAndCreate(tartgetFile)) {
            //若目标文件创建失败
            return false;
        }

        InputStream inputStream = null;
        BufferedOutputStream bufferedOutputStream = null;
        try {
            inputStream = new FileInputStream(sourceFile);
            bufferedOutputStream =
                    new BufferedOutputStream(new FileOutputStream(tartgetFile, false));
            byte data[] = new byte[BUFFER_SIZE];
            int lengh;
            while ((lengh = inputStream.read(data, 0, BUFFER_SIZE)) != -1) {
                bufferedOutputStream.write(data, 0, lengh);
            }
            return true;
        } catch (IOException e) {
            e.printStackTrace();
            return false;
        } finally {
            if (inputStream != null) {
                closeIO(inputStream);
            }
            if (bufferedOutputStream != null) {
                closeIO(bufferedOutputStream);
            }
        }
    }
}

File: utilsLibrary/src/main/java/com/zeropercenthappy/utilslibrary/utils/FileUtils.kt
package com.zeropercenthappy.utilslibrary.utils

import android.webkit.MimeTypeMap
import java.io.*
import java.util.*

object FileUtils {


    private const val BUFFER_SIZE = 8192

    @JvmStatic
    fun closeIO(vararg closeables: Closeable?) {
        if (closeables.isNullOrEmpty()) {
            return
        }
        for (io in closeables) {
            try {
                io?.run { close() }
            } catch (exception: IOException) {
                exception.printStackTrace()
            }
        }
    }

    /**
     * Check the file is exist or not. If the file isn't exist, it will try to create all needs
     * directory and itself.
     *
     * @param file  destination file
     *
     * @return      true or false(if not exist and create failure)
     */
    @JvmStatic
    fun checkFileAndCreate(file: File): Boolean {
        return try {
            val dir = file.parentFile
            dir.mkdirs()
            if (file.exists()) {
                true
            } else {
                file.createNewFile()
                file.exists()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    /**
     * Check the file is exist or not. If the file isn't exist, it will try to create all needs
     * directory and itself.
     *
     * @param path  destination file's absolute path
     *
     * @return      true or false(if not exist and create failure)
     */
    @JvmStatic
    fun checkFileAndCreate(path: String): Boolean {
        return try {
            val file = File(path)
            val dir = file.parentFile
            dir.mkdirs()
            if (file.exists()) {
                true
            } else {
                file.createNewFile()
                file.exists()
            }
        } catch (e: Exception) {
            e.printStackTrace()
            false
        }
    }

    /**
     * Write file through InputStream.
     *
     * @param storageFile   storageFile
     *
     * @param inputStream   inputStream
     *
     * @return              write result
     */
    @JvmStatic
    fun writeFileByIS(storageFile: File, inputStream: InputStream): Boolean {
        var bufferedOutputStream: BufferedOutputStream? = null
        try {
            if (!checkFileAndCreate(storageFile)) {
                return false
            }
            bufferedOutputStream = BufferedOutputStream(FileOutputStream(storageFile))
            val buffer = ByteArray(BUFFER_SIZE)
            var currentLength: Int
            while (true) {
                currentLength = inputStream.read(buffer, 0, BUFFER_SIZE)
                if (currentLength != -1) {
                    bufferedOutputStream.write(buffer, 0, currentLength)
                } else {
                    break
                }
            }
            bufferedOutputStream.flush()
            return true
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        } finally {
            closeIO(inputStream)
            closeIO(bufferedOutputStream)
        }
    }

    /**
     * Write string to a file.
     *
     * @param file      file
     *
     * @param content   content
     *
     * @param append    is append
     *
     * @return          write result
     */
    @JvmStatic
    fun writeString2File(file: File, content: String, append: Boolean) = try {
        val bufferedWriter = BufferedWriter(FileWriter(file.absolutePath, append))
        bufferedWriter.write(content)
        bufferedWriter.flush()
        closeIO(bufferedWriter)
        true
    } catch (e: IOException) {
        e.printStackTrace()
        false
    }


    /**
     *  Get file's MimeType.
     *
     * @param file  Destination file
     *
     * @return      file's MimeType or null
     */
    @JvmStatic
    fun getFileMimeType(file: File): String? {
        return getFileMimeTypeByExtension(file.extension)
    }

    /**
     * Get file's MimeType.
     *
     * @param path  File's absolute path
     *
     * @return      file's MimeType or null
     */
    @JvmStatic
    fun getFileMimeType(path: String): String? {
        val file = File(path)
        return getFileMimeTypeByExtension(file.extension)
    }

    /**
     * Get file's MimeType by extension.
     *
     * @param extension 后缀名
     *
     * @return          file's MimeType or null
     */
    @JvmStatic
    fun getFileMimeTypeByExtension(extension: String): String? {
        return MimeTypeMap
            .getSingleton()
            .getMimeTypeFromExtension(extension.toLowerCase(Locale.getDefault()))
    }

    /**
     * Get file's size, unit：Byte
     *
     * @param file  file
     *
     * @return      file's size
     */
    fun getSize(file: File?): Long {
        var size: Long = 0
        if (file == null || !file.exists()) {
            return size
        }
        if (file.isFile) {
            return file.length()
        } else if (file.isDirectory) {
            val files = file.listFiles()
            for (subFile in files) {
                size += getSize(subFile)
            }
        }
        return size
    }

    /**
     * Copy file.
     *
     * @param sourceFile  source file
     *
     * @param destinationFile  destination file
     *
     * @return copy result
     */
    fun copyFile(sourceFile: File, destinationFile: File): Boolean {
        if (!sourceFile.exists()) {
            // 若源文件不存在
            return false
        }
        if (!deleteFile(destinationFile)) {
            // 若旧目标文件删除失败
            return false
        }
        if (!checkFileAndCreate(destinationFile)) {
            // 若目标文件创建失败
            return false
        }
        var inputStream: InputStream? = null
        var bufferedOutputStream: BufferedOutputStream? = null
        try {
            inputStream = FileInputStream(sourceFile)
            bufferedOutputStream = BufferedOutputStream(FileOutputStream(destinationFile, false))
            val data = ByteArray(BUFFER_SIZE)
            while (true) {
                val length = inputStream.read(data, 0, BUFFER_SIZE)
                if (length != -1) {
                    bufferedOutputStream.write(data, 0, length)
                } else {
                    break
                }
            }
            return true
        } catch (e: IOException) {
            e.printStackTrace()
            return false
        } finally {
            if (inputStream != null) {
                closeIO(inputStream)
            }
            if (bufferedOutputStream != null) {
                closeIO(bufferedOutputStream)
            }
        }
    }

    /**
     * Delete file or directory
     *
     * @param file  destination file or directory
     *
     * @return      delete result
     */
    @JvmStatic
    fun deleteFile(file: File): Boolean {
        if (!file.exists()) {
            return false
        }
        return when {
            file.isFile -> file.delete()
            file.isDirectory -> {
                val files = file.listFiles()
                files?.run {
                    for (childFile in this) {
                        deleteFile(childFile)
                    }
                }
                file.delete()
            }
            else -> false
        }
    }

}